loadtest/loadtest.go

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package loadtest
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/mattermost/mattermost-load-test-ng/config"
+	"github.com/mattermost/mattermost-load-test-ng/loadtest/control"
+
+	"github.com/mattermost/mattermost-server/v5/mlog"
+)
+
+// LoadTester is a structure holding all the state needed to run a load-test
+type LoadTester struct {
+	controllers   []control.UserController
+	config        *config.LoadTestConfig
+	wg            sync.WaitGroup
+	status        chan control.UserStatus
+	started       bool
+	newController NewController
```

Comment:
Is there a reason we are using this factory function here ? Is it just to de-duplicate the initialization code ?

Reply:
The idea around the factory function is that we don't want the `LoadTester` to know anything about how the users/controllers are implemented so that the same code will work for any implementation of those two interfaces.  
At the same time we want to be able to create the concrete values from the inside without requiring the caller to provide them to us, so as to remove some complexity from the user's side. I find passing a factory function to be an elegant solution in this case.  
The reason the function is stored inside `LoadTester` is because we pass it once during initialization and then we use it at later time in `AddUser`.

Reply:
Ah okay, so you are saying there will be different user/controller implementations, and it will be cumbersome to change an internal `LoadTester` method every time we want to switch to a different implementation ?

I think this is a great point and should be noted in the godoc for `NewContoller`.

Reply:
Yes, there will definitely be at least two `UserController` implementations, the *simple* one we are building now and a more *simulative* one yet to be coded.  
I'll improve the documentation including some of these thoughts so to make clear the reasoning behind these design choices.